akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -40,38 +39,53 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with StageLoggi
         if (streamId <= largestIncomingStreamId) Closed // closed streams are never put into the map
         else {
           largestIncomingStreamId = streamId
-          incomingStreams += streamId -> Idle
-          Idle
+          val initialState = Idle(outgoingStreamEnded = false)
+          incomingStreams += streamId -> initialState
+          initialState
         }
     }
   def handleStreamEvent(e: StreamFrameEvent): Unit = {
-    val newState = streamFor(e.streamId).handle(e)
-    if (newState == Closed) incomingStreams -= e.streamId
-    else incomingStreams += e.streamId -> newState
+    updateState(e.streamId, streamFor(e.streamId).handle(e))
+  }
+  def handleOutgoingEnded(streamId: Int): Unit = {
+    updateState(streamId, streamFor(streamId).handleOutgoingEnded())
+  }
+  def updateState(streamId: Int, newState: IncomingStreamState): Unit = {
+    if (newState == Closed) incomingStreams -= streamId
+    else incomingStreams += streamId -> newState
   }
   def resetStream(streamId: Int, errorCode: ErrorCode): Unit = {
-    // FIXME: put this stream into an extra state where we allow some frames still to be received
     incomingStreams -= streamId
     multiplexer.pushControlFrame(RstStreamFrame(streamId, errorCode))
   }
 
   sealed abstract class IncomingStreamState { _: Product =>
     def handle(event: StreamFrameEvent): IncomingStreamState
+    def handleOutgoingEnded(): IncomingStreamState
 
     def stateName: String = productPrefix
     def receivedUnexpectedFrame(e: StreamFrameEvent): IncomingStreamState = {
       pushGOAWAY(ErrorCode.PROTOCOL_ERROR, s"Received unexpected frame of type ${e.frameTypeName} for stream ${e.streamId} in state $stateName")
       Closed
     }
   }
-  case object Idle extends IncomingStreamState {
+  case class Idle(outgoingStreamEnded: Boolean) extends IncomingStreamState {
```

Comment:
(perhaps we should make this 2 objects to reduce garbage - might not be worth it though?)

Reply:
These states are meant to correspond to the states in the [state diagram](https://httpwg.org/specs/rfc7540.html#StreamStates) of the RFC (we should document that).

Wrt that it's not obvious to me what the new state is and how we would end up there? I'd say we should treat it as an error.

Reply:
I guess the HTTP/2 terminology might be a bit misleading in that regard. "Idle" means "so far unused" instead of "used before but currently unused".

Another confusing thing might be, that - so far - the logic is only for the server side. On the server side, the only way (leaving `PUSH_PROMISE` aside for now) to open a stream is receiving a `HEADERS` frames, so there has been no traffic in the outgoing direction at all so far.

On the client side there will be another edge in the state graph. In that case, we will move from `Idle` to `Open` when we, the client, send out our original `HEADERS` frame. Then we might (and will in a standard request/response case) indeed see an incoming `HEADERS` frame (= start of response) after the outgoing stream (= request stream) has been closed.  So, we can still keep the simple Idle state but will have to introduce another handler for moving to `Open` when we create a new stream actively.

Reply:
I agree on closer examination that this was overly defensive.

In the state diagram, there is no way we could send `EndStream` in the `Idle` state: after all, we haven't received headers so the outgoing stream can't have started. Since in our implementation we have separated the incoming and outgoing streams, it wasn't entirely obvious to me there couldn't be a race condition where we'd see the `sending EndStream` before we saw the `incoming Headers`. Looking at it a bit better, though, since we do `dispatchSubstream` inside `Idle` message handling, this is actually safe and we don't need the separate flag to remember we've sent the `EndStream` even though we were still in the `Idle` state.